mountpoint-s3/src/prefetch.rs

```diff
@@ -347,7 +386,17 @@ where
                 .unwrap();
 
             self.next_sequential_read_offset += part_bytes.len() as u64;
-            self.prepare_requests();
+
+            // Increase the read window size if available bytes in the queue are less than half of previous read window size
+            let remaining_in_queue = current_task
+                .available_offset()
+                .saturating_sub(self.next_sequential_read_offset) as usize;
+            if remaining_in_queue < self.preferred_read_window_size / 2 {
+                self.preferred_read_window_size = (self.preferred_read_window_size
```

Comment:
Do we need `self.preferred_read_window_size` now? Maybe it should belong to `RequestTask`, which could handle the read window increase when triggered here.

Reply:
I'm not sure where it should be either, but I believe the prefetcher should be in control of scaling up or down of the read window size. That's why I keep it here.

Reply:
+1 to Alex, I'd say this logic belongs to `RequestTask` (or even to `BackpressureLimiter`?), in that way window-moving logic won't be scattered around the code base and there will be no necessity for `RequestTask` to look inside `RequestTask` for `available_offset`.

by the way,  wouldn't a simplified interface like this suffice for `BackpressureController`?
```rust
impl BackpressureController {
    // Used by `PrefetchGetObject` to report to `BackpressureController` the offset 
    // until which the data was read by the user application.
    pub fn set_next_sequential_read_offset(&mut self, next_sequential_read_offset: usize)
}
```

Ultimately `next_sequential_read_offset` seems to be the only bit of information, which `PrefetchGetObject` should report to `BackpressureLimiter` (via `BackpressureController`) so that the later can make a decision to move the window (apart from the information provided on `RequestTask` instantiation: `preferred_part_size`, `object_size`).

Reply:
the whole thing becomes much clearer than:
1. we have a `RequestTask` pumping data from CRT to the queue;
2. we have a `PrefetchGetObject` reading that data from the queue when user requests it;
3. `RequestTask` and `PrefetchGetObject` are synchronised through an additional queue, which is used by `PrefetchGetObject` to report when the data is consumed;
4. `RequestTask` stops pumping next data parts from CRT (in `next_read_window_blocking` call) until `PrefetchGetObject` consumes something from the previous part (and reports `next_sequential_read_offset`)

Reply:
Just to clarify, the logic here is not for moving the read window but for deciding if we should use bigger window size. Agree that the window moving logic should be in the `RequestTask`.

Reply:
Looks overcomplicated when we implement window calculation in `PrefetchGetObject`. As I mentioned it needs to fetch an atomic `available_offset` from the `RequestTask` and pass the new window size through the channel to `ClientPartStream`, while it could've passed more basic information in form of consumed offset.

But if you have a strong opinion of why window calculation logic should be in `PrefetchGetObject`, please, share your reasoning!

Reply:
Not a really strong opinion, if the logic is in `RequestTask` we still have to pass information about how to scale up the request (like `max_read_window_size` or `sequential_prefetch_multiplier`) to the task as they are currently in the `PrefetcherConfig`. I'm also thinking about if we have some kind of memory limiter in the future, we will also have to push the information/logic about how to scale down to the task.

Reply:
Ack, some of the information originates from `PrefetcherConfig` and we want to keep it there.